framework/fp_vectors/Floats.h

```diff
@@ -223,66 +295,202 @@ struct BFloat16
-    double as_float;
-    uint64_t as_hex;
-    struct {
-        unsigned long long mantissa : 52;
-        unsigned long long exponent : 11;
-        unsigned long long sign : 1;
+
+#ifdef __cplusplus
+    inline Float32() = default;
+    constexpr inline Float32(float f) : as_float(f) { }
+    constexpr inline Float32(uint64_t m, uint32_t e, uint32_t s): mantissa(m), exponent(e), sign(s) { }
+#endif
+};
+typedef struct Float32 Float32;
+
+struct Float64 {
+    union {
+        struct {
+            uint64_t mantissa : FLOAT64_MANTISSA_BITS;
+            uint64_t exponent : FLOAT64_EXPONENT_BITS;
+            uint64_t sign     : FLOAT64_SIGN_BITS;
+        };
+        struct {
+            uint32_t low32;
+            uint32_t high32;
+        };
+        double as_float;
+        uint64_t as_hex;
     };
-} Float64;
-static_assert(sizeof(Float64) == sizeof(double), "Float64 structure is not of the correct size");
 
-typedef union {
-    struct {
+#ifdef __cplusplus
+    inline Float64() = default;
+    constexpr inline Float64(float f) : as_float(f) { }
+    constexpr inline Float64(uint64_t m, uint32_t e, uint32_t s): mantissa(m), exponent(e), sign(s) { }
+#endif
+};
+typedef struct Float64 Float64;
+
+struct Float80 {
+    union {
+        struct {
+            uint64_t mantissa : FLOAT80_MANTISSA_BITS;
+            uint64_t jbit     : FLOAT80_JBIT_BITS;
+            uint64_t exponent : FLOAT80_EXPONENT_BITS;
+            uint64_t sign     : FLOAT80_SIGN_BITS;
+        };
+        struct {
+            uint64_t low64;
+            uint16_t high16;
+        } as_hex;
+        struct {
+            uint32_t low32;
+            uint32_t high32;
+            uint16_t extra16;
+        };
         long double as_float;
     };
-    struct {
-        unsigned long long mantissa : 63;
-        unsigned long jbit : 1;
-        unsigned exponent : 15;
-        unsigned sign : 1;
-    };
-    struct {
-        uint64_t low64;
-        uint16_t high16;
-    } as_hex;
-} Float80;
-static_assert(sizeof(double) < sizeof(long double), "Compiler does not support long double");
-static_assert(sizeof(Float80) == sizeof(long double), "Float80 structure is not of the correct size");
-
-Float16 new_float16(unsigned sign, unsigned exponent, unsigned mantissa);
-Float32 new_float32(uint32_t sign, uint32_t exponent, uint32_t mantissa);
-Float64 new_float64(uint32_t sign, uint32_t exponent, uint64_t mantissa);
-Float80 new_float80(uint32_t sign, uint32_t exponent, uint32_t jbit, uint64_t mantissa);
+
+#ifdef __cplusplus
+    inline Float80() = default;
+    constexpr inline Float80(long double f) : as_float(f) { }
+    constexpr inline Float80(uint64_t m, uint32_t j, uint32_t e, uint32_t s): mantissa(m), jbit(j), exponent(e), sign(s) { }
+#endif
+};
+typedef struct Float80 Float80;
+
+/**
+ * @brief C/C++ builders (inlined)
+ *
+ *
+ * @{
+ */
+#ifdef __cplusplus
+#define STATIC_INLINE static inline constexpr
+#else
+#define STATIC_INLINE static inline
+#endif
```

Comment:
I feel a bit uneasy about this macro leaking outside this header.
It looks like framework-wide used macro, but defined and used only inside Floats.h.

Maybe we should do:
```cpp
#ifdef __cplusplus
#define SANDSTONE_CONSTEXPR constexpr
#else
#define SANDSTONE_CONSTEXPR
#endif
```
and move it to `sandstone_data.h`.

Also, keep in mind that `inline` specifier is different in C and C++.
Maybe we need something like `#define SANDSTONE_FORCE_INLINE __attribute__((always_inline))`.

btw macro like that will soon be added to one of the framework headers.